### ISSUE TITLE
Return reason as error when `may_...?` is invoked

### ISSUE THREAD

**Issue by nitsujri** [aasm/aasm#676]
**Is your feature request related to a problem? Please describe.**
Automatically add failure to transition reason into `object.errors`.

**Describe the solution you'd like**
``` ruby
class Transaction
  aasm do
    state :verifying
    state :completed
    state :cancelled

    event :cancel, guard: :able_to_cancel? do
      transitions to: :cancelled, from: :verifying
    end
  end

  def able_to_cancel?
    return true unless money.deposited?

    errors.add(:aasm_state, 'Money was deposited, so we cannot cancel')
    false
  end
end

# else_where.rb
some_transaction.aasm_state # verifying
some_transaction.may_cancel?
some_transaction.errors.full_messages # Contains only error of money deposited

some_transaction.aasm_state # completed
some_transaction.may_cancel?
some_transaction.errors.full_messages # Contains both errors of transition & money deposited
```

**Describe alternatives you've considered**
It would be nice to tell the user why something cannot transition without attempting the transition itself.

It is possible to do a non-committal `some_transaction.cancel` but that is attempting the transition which blurs the line of performing action instead of only informing.

In the example above, we rarely want to cancel a transaction but it's good to know what are the reasons you cannot do something instead of simply blocking without explanation.

Thanks for this great gem!

**Comment by anilmaurya** [comment]
Hi @nitsujri 

Scenaio 1
```
some_transaction.aasm_state # verifying
some_transaction.may_cancel?
some_transaction.errors.full_messages # Contains error of money deposited
```

Scenario 2
```
some_transaction.aasm_state # completed
some_transaction.may_cancel?
some_transaction.errors.full_messages # {}
```
Guard errors are added to `object.errors`, therefore scenario 1  is working as expected.

Transition error is not added to `object.errors` therefore scenario 2 shows empty error.

Adding transition error to `object.errors` is possible but adding guards error from some other transition will be tricky.

WDYT ?

**Comment by nitsujri** [comment]
> Adding transition error to object.errors is possible but adding guards error from some other transition will be tricky.

For clarification, do you mean that adding guard errors from the *same* transition will be tricky?

If it's difficult to run both the guard and the transition such that they both add to `errors` I can understand that.

If you're saying "*from some other transition will be tricky*", I don't fully grasp. I'm not really looking to run a different transition other than the one outlined in the `cancel` event.

I'm probably missing something, so let me know. Thanks!

**Comment by anilmaurya** [comment]
Hi @nitsujri

**Comment by nitsujri** [comment]
> Moving from `completed` to `cancelled` , you want money deposited error but money deposited error is added in guard for transition from `verifying` to `cancelled` therefore I thought you are looking for errors from other transition as well.

Ah this is where my knowledge of AASM's internal workings is lacking. 

I assumed that since the guard `able_to_cancel?` is declared at the "event level" it would run for all transition cases simply because I called `.cancel`/`.may_cancel?` (specifically as opposed to being declared at the "transition level" using `transition to: :some_state, guard: :my_guard`). 

What you're telling me is, guards only run for explicitly valid transition cases?

If that's the case:

- Base Feature Request: Add error responses for blocking/failing transitions.
- 2nd Feature Request: Run "event level" guards at every event call along with transitions.

The 2nd request is this is likely a breaking change and a heavy one at that. I would not necessarily believe it is worth the effort, but I will say that's what I would expect to have happen by reading the code at face value (i.e. it guards the `event` instead of the `transition`?).

**Comment by anilmaurya** [comment]
> What you're telling me is, guards only run for explicitly valid transition cases?

Yes, that's correct. Guard run for explicitly valid transition cases

So, AASM expects all valid transitions to be defined.

If you define event like below:
```
event :cancel, guard: :able_to_cancel? do
      transitions to: :cancelled, from: :verifying, guard: :some_other_validation
      transitions to: :cancelled, from: :completed
end
```
Then it will run `:able_to_cancel?` guard for both transitions but `:some_other_validation` guard will only get execute when transitioning from :verifying to :cancelled state

**Comment by nitsujri** [comment]
@anilmaurya thanks that makes a lot of sense.

How would you like me to handle this feature request? Split the base/2nd request cases? Close as won't fix? 

I still think that the base case of adding transition blockers to `.errors` is valid, but only if you believe so as well.

**Comment by anilmaurya** [comment]
Yes, 2nd Feature request can be marked as won't fix.

I agree that adding transition blockers to `errors` is valid use case.
